### DISCUSSION TITLE
CMake + Ccache: two popular approaches

### DISCUSSION THREAD

**Discussion by diiigle** [AcademySoftwareFoundation/openvdb#1292] [General]
In spirit of this [stackoverflow question](https://stackoverflow.com/questions/64592095/cmake-ccache-rule-launch-compile-or-cmake-lang-compiler-launcher), I want to raise awareness about how Ccache is activated in openvdb:

There seem to be two approaches, with only a slight tendency to the second one (CMAKE_*_COMPILER_LAUNCHER).
You can select it per language, leave users the freedom to overwrite it with ENV variables and change it per directory + target. 
[This book](https://cliutils.gitlab.io/modern-cmake/chapters/features/utilities.html) also recommends the second one.

 If you have a parent project using this second approach, for some reason you end up with nested `ccache ccache` calls (which fail).

My current workaround is to `set(USE_CCACHE OFF)` for openvdb to avoid the recursive calls. But it would be cleaner to have openvdb also use the compatible approach. I can prepare a PR after a discussion, but it should be only changing:
https://github.com/AcademySoftwareFoundation/openvdb/blob/5ecc250302807c3854e05f1f6a0ea6bf7daf29bb/CMakeLists.txt#L404-L405

**Comment by theHamsta** [comment]
Why not use 
```
export CC=/usr/lib/ccache/clang-12
export CXX=/usr/lib/ccache/clang++-12
```
I don't see a reason why the built system should handle the usage of ccache at all. Of course you can also use the `CMAKE_*_COMPILER_LAUNCHER` approach, but does this require having ccache in the CMake script at all?

**Comment by diiigle** [comment]
If I would need to guess, I'd say the reason CMake is involved, is convenience. You want to give new developers / users the chance to use ccache, but not enforce it. For experienced ccache users like you, of course you can use the manual injection-from-the-outside method.

To my knowledge, the whole reason cmake implemented these `CMAKE_*_COMPILER_LAUNCHER` variables is caches like ccache.  It probably exists to reduce the manual labour setting up the environment around a project. 

In the end I don't want to discuss WHY it is there, but HOW. Thanks for pointing out yet another option.

As for your question, you can also seed this variable from an environment variable, without support of the cmake script. Again, it's convenience why you let cmake look for ccache automatically and enable it when found.

**Comment by theHamsta** [comment]
But you can use `CMAKE_*_COMPILER_LAUNCHER` without having them in the CMake script.
